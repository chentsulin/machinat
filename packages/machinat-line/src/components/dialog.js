/* eslint-disable import/prefer-default-export */
import { annotateLineComponent, isMessageValue } from '../utils';

export const Dialog = async (
  { props: { children, quickReplies } },
  path,
  render
) => {
  const segments = await render(children, '.children');
  if (segments === null) {
    return null;
  }

  let lastMessageIdx = -1;
  // hoisting text to line text message object
  for (let i = 0; i < segments.length; i += 1) {
    const segment = segments[i];
    const { type, value } = segment;

    if (type === 'text') {
      segment.type = 'unit';
      segment.value = {
        type: 'text',
        text: value,
      };
    }

    if (isMessageValue(value)) {
      lastMessageIdx = i;
    }
  }

  const quickReplySegments = await render(quickReplies, '.quickReplies');

  if (quickReplySegments) {
    if (lastMessageIdx === -1) {
      throw new Error('no message existed in children to attach quickReply');
    }

    segments[lastMessageIdx].value.quickReply = {
      items: quickReplySegments.map(segment => segment.value),
    };
  }

  return segments;
};
annotateLineComponent(Dialog);
