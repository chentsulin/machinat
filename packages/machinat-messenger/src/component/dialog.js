/* eslint-disable import/prefer-default-export */
import valuesOfAssertedTypes from '@machinat/core/utils/valuesOfAssertedTypes';

import { asContainerComponent, isMessageEntry } from '../utils';
import * as quickReplyModule from './quickReply';

const getQuickRepliesValues = valuesOfAssertedTypes(() => [
  ...Object.values(quickReplyModule),
]);

const Dialog = async (
  { children, type, tag, notificationType, metadata, quickReplies, personaId },
  render
) => {
  const childrenSegments = await render(children, '.children');
  if (childrenSegments === null) {
    return null;
  }

  let lastMessageIdx = -1;

  for (let i = 0; i < childrenSegments.length; i += 1) {
    const segment = childrenSegments[i];

    // hoisting text to message object
    if (typeof segment.value === 'string') {
      segment.type = 'unit';
      segment.value = {
        message: {
          text: segment.value,
        },
      };
    }

    const { value } = segment;

    if (isMessageEntry(value)) {
      const copied = { ...value };

      if (value.message) {
        copied.messaging_type = type;
        copied.tag = tag;
        copied.notification_type = notificationType;
        copied.persona_id = personaId;

        lastMessageIdx = i;
      } else if (
        value.sender_action === 'typing_on' ||
        value.sender_action === 'typing_off'
      ) {
        copied.persona_id = personaId;
      }

      segment.value = copied;
    }
  }

  // only attach quick_replies and metadata to last message segment
  if (lastMessageIdx !== -1) {
    const { value } = childrenSegments[lastMessageIdx];

    const message = { ...value.message };
    message.metadata = metadata;

    const quickReplySegments = await render(quickReplies, '.quickReplies');
    message.quick_replies = getQuickRepliesValues(quickReplySegments);

    value.message = message;
  }

  return childrenSegments;
};

const __Dialog = asContainerComponent(Dialog);

export { __Dialog as Dialog };
