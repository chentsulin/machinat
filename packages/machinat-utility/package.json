{
  "name": "machinat-utility",
  "version": "0.0.1-4",
  "author": "Patrick Liu <zeldalink0515@gmail.com>  (https://github.com/lrills)",
  "repository": {
    "type": "git",
    "url": "https://github.com/lrills/machinat.git",
    "directory": "packages/machinat-utility"
  },
  "homepage": "https://github.com/lrills/machinat#readme",
  "bugs": "https://github.com/lrills/machinat",
  "main": "lib/index.js",
  "files": [
    "lib/*",
    "README.md"
  ],
  "license": "MIT",
  "dependencies": {
    "invariant": "^2.2.4"
  },
  "peerDependencies": {
    "machinat": "^0.0.1-1"
  }
}
