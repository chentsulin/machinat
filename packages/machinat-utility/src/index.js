// @flow
export { default as compose } from './compose';
export { default as isValidElementType } from './isValidElementType';
export { default as formatNode } from './formatNode';
export { default as joinTextualSegments } from './joinTextualSegments';
export { default as valuesOfAssertedType } from './valuesOfAssertedType';
export { default as mixin } from './mixin';
export { default as toJSONWithProto } from './toJSONWithProto';

export * from './isX';
export * from './children';
