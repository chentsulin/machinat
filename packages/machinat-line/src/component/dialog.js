/* eslint-disable import/prefer-default-export */
import { valuesOfAssertedType } from 'machinat-utility';
import { asContainerComponent } from './utils';
import { QuickReply } from './quickReply';

const renderQuickReplyValues = valuesOfAssertedType(QuickReply);

const Dialog = ({ props: { children, quickReplies } }, render) => {
  const segments = render(children, '.children');
  if (segments === null) {
    return null;
  }

  let lastMessageIdx = -1;
  // hoisting text to line text message object
  for (let i = 0; i < segments.length; i += 1) {
    const segment = segments[i];
    const { type, node, value } = segment;

    if (type === 'text') {
      segment.type = 'unit';
      segment.value = {
        type: 'text',
        text: value,
      };
    }

    if (
      typeof node !== 'object' ||
      typeof node.type.$$getEntry !== 'function'
    ) {
      lastMessageIdx = i;
    }
  }

  const quickRepliesValues = renderQuickReplyValues(
    render(quickReplies, '.quickReplies')
  );

  if (quickRepliesValues && lastMessageIdx !== -1) {
    segments[lastMessageIdx].value.quickReply = {
      items: quickRepliesValues,
    };
  }

  return segments;
};

const __Dialog = asContainerComponent(Dialog);

export { __Dialog as Dialog };
