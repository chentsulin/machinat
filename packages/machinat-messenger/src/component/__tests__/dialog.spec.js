import Machinat from 'machinat';
import { map } from 'machinat-utility';

import {
  QuickReply,
  LocationQuickReply,
  EmailQuickReply,
  PhoneQuickReply,
} from '../quickReply';

import { Dialog } from '../dialog';

import renderHelper from './renderHelper';

const quickReplies = (
  <>
    <QuickReply title="foo" />
    <LocationQuickReply />
    <PhoneQuickReply />
    <EmailQuickReply />
  </>
);

const children = (
  <>
    foo
    <bar />
  </>
);

const Unknown = () => {};

const renderInner = jest.fn(
  message =>
    map(
      message,
      (node, path) =>
        typeof node === 'string'
          ? { type: 'unit', value: node, node, path }
          : typeof node.type === 'string'
          ? {
              type: 'unit',
              value: { message: { text: node.type } },
              node,
              path,
            }
          : node.type === Unknown
          ? { type: 'unit', value: { you: "don't know me" }, node, path }
          : { type: 'unit', value: '_QUICK_REPLY_RENDERED_', node, path },
      '$'
    ) || null
);

const render = renderHelper(renderInner);

it.each([
  ['Dialog with type', <Dialog type="UPDATE">{children}</Dialog>],
  [
    'Dialog with type and tag',
    <Dialog type="MESSAGE_TAG" tag="COMMUNITY_ALERT">
      {children}
    </Dialog>,
  ],
  [
    'Dialog with notificationType',
    <Dialog notificationType="NO_PUSH">{children}</Dialog>,
  ],
  [
    'Dialog with metadata',
    <Dialog metadata="_SOME_METADATA_">{children}</Dialog>,
  ],
  [
    'Dialog with quickReplies',
    <Dialog quickReplies={quickReplies}>{children}</Dialog>,
  ],
  [
    'Dialog with personaId',
    <Dialog personaId="_PERSONA_ID_">{children}</Dialog>,
  ],
  [
    'Dialog with all props',
    <Dialog
      type="MESSAGE_TAG"
      tag="PAYMENT_UPDATE"
      notificationType="SILENT_PUSH"
      metadata="_MY_METADATA_"
      quickReplies={quickReplies}
      personaId="_PERSONA_ID_"
    >
      {children}
    </Dialog>,
  ],
])('%s match snapshot', (_, element) => {
  expect(render(element)).toMatchSnapshot();
});

it('returns null if empty children received', () => {
  expect(render(<Dialog />)).toBe(null);
  expect(render(<Dialog>{null}</Dialog>)).toBe(null);
});

it('hoist text value into message object', () => {
  const actions = render(
    <Dialog>
      foo
      <bar />
      baz
    </Dialog>
  );

  expect(actions[0].value).toEqual({ message: { text: 'foo' } });
  expect(actions[1].value).toEqual({ message: { text: 'bar' } });
  expect(actions[2].value).toEqual({ message: { text: 'baz' } });
});

it('add root fields to action value', () => {
  const segments = render(
    <Dialog
      type="MESSAGE_TAG"
      tag="PAYMENT_UPDATE"
      notificationType="SILENT_PUSH"
      personaId="_PERSONA_ID_"
    >
      <foo />
      <bar />
    </Dialog>
  );

  segments.forEach(segment => {
    expect(segment.value.messaging_type).toBe('MESSAGE_TAG');
    expect(segment.value.tag).toBe('PAYMENT_UPDATE');
    expect(segment.value.notification_type).toBe('SILENT_PUSH');
    expect(segment.value.persona_id).toBe('_PERSONA_ID_');
  });
});

it('adds metadata to last message action', () => {
  const actions = render(
    <Dialog metadata="_META_">
      <foo />
      bar
      <Unknown />
    </Dialog>
  );

  expect(actions[0].value).toEqual({ message: { text: 'foo' } });
  expect(actions[1].value).toEqual({
    message: { text: 'bar', metadata: '_META_' },
  });
  expect(actions[2].value).toEqual({ you: "don't know me" });
});

it('adds quickReplies to last message action', () => {
  const actions = render(
    <Dialog
      quickReplies={[
        <QuickReply title="foo" />,
        <LocationQuickReply />,
        <PhoneQuickReply />,
        <EmailQuickReply />,
      ]}
    >
      <foo />
      bar
      <Unknown />
    </Dialog>
  );

  expect(actions[0].value).toEqual({ message: { text: 'foo' } });
  expect(actions[1].value).toEqual({
    message: {
      text: 'bar',
      quick_replies: [
        '_QUICK_REPLY_RENDERED_',
        '_QUICK_REPLY_RENDERED_',
        '_QUICK_REPLY_RENDERED_',
        '_QUICK_REPLY_RENDERED_',
      ],
    },
  });
  expect(actions[2].value).toEqual({ you: "don't know me" });
});

it('throw if non QuickReply element received within prop quickReplies', () => {
  expect(() =>
    render(
      <Dialog
        quickReplies={[
          <QuickReply title="foo" />,
          <LocationQuickReply />,
          <Unknown />,
          <PhoneQuickReply />,
          <EmailQuickReply />,
        ]}
      >
        <foo />
        bar
      </Dialog>
    )
  ).toThrowErrorMatchingInlineSnapshot(
    `"<Unknown /> at $:2 is invalid, only <[QuickReply, LocationQuickReply, PhoneQuickReply, EmailQuickReply]/> allowed"`
  );
});
