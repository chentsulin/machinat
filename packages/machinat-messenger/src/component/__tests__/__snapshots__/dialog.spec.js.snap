// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dialog with all props match snapshot 1`] = `
Array [
  Object {
    "node": "foo",
    "path": "$::0",
    "type": "unit",
    "value": Object {
      "message": Object {
        "text": "foo",
      },
      "messaging_type": "MESSAGE_TAG",
      "notification_type": "SILENT_PUSH",
      "persona_id": "_PERSONA_ID_",
      "tag": "PAYMENT_UPDATE",
    },
  },
  Object {
    "node": Object {
      "$$typeof": Symbol(machinat.element),
      "props": Object {},
      "type": "bar",
    },
    "path": "$::1",
    "type": "unit",
    "value": Object {
      "message": Object {
        "metadata": "_MY_METADATA_",
        "quick_replies": Array [
          "_QUICK_REPLY_RENDERED_",
          "_QUICK_REPLY_RENDERED_",
          "_QUICK_REPLY_RENDERED_",
          "_QUICK_REPLY_RENDERED_",
        ],
        "text": "bar",
      },
      "messaging_type": "MESSAGE_TAG",
      "notification_type": "SILENT_PUSH",
      "persona_id": "_PERSONA_ID_",
      "tag": "PAYMENT_UPDATE",
    },
  },
]
`;

exports[`Dialog with metadata match snapshot 1`] = `
Array [
  Object {
    "node": "foo",
    "path": "$::0",
    "type": "unit",
    "value": Object {
      "message": Object {
        "text": "foo",
      },
      "messaging_type": undefined,
      "notification_type": undefined,
      "persona_id": undefined,
      "tag": undefined,
    },
  },
  Object {
    "node": Object {
      "$$typeof": Symbol(machinat.element),
      "props": Object {},
      "type": "bar",
    },
    "path": "$::1",
    "type": "unit",
    "value": Object {
      "message": Object {
        "metadata": "_SOME_METADATA_",
        "quick_replies": undefined,
        "text": "bar",
      },
      "messaging_type": undefined,
      "notification_type": undefined,
      "persona_id": undefined,
      "tag": undefined,
    },
  },
]
`;

exports[`Dialog with notificationType match snapshot 1`] = `
Array [
  Object {
    "node": "foo",
    "path": "$::0",
    "type": "unit",
    "value": Object {
      "message": Object {
        "text": "foo",
      },
      "messaging_type": undefined,
      "notification_type": "NO_PUSH",
      "persona_id": undefined,
      "tag": undefined,
    },
  },
  Object {
    "node": Object {
      "$$typeof": Symbol(machinat.element),
      "props": Object {},
      "type": "bar",
    },
    "path": "$::1",
    "type": "unit",
    "value": Object {
      "message": Object {
        "metadata": undefined,
        "quick_replies": undefined,
        "text": "bar",
      },
      "messaging_type": undefined,
      "notification_type": "NO_PUSH",
      "persona_id": undefined,
      "tag": undefined,
    },
  },
]
`;

exports[`Dialog with personaId match snapshot 1`] = `
Array [
  Object {
    "node": "foo",
    "path": "$::0",
    "type": "unit",
    "value": Object {
      "message": Object {
        "text": "foo",
      },
      "messaging_type": undefined,
      "notification_type": undefined,
      "persona_id": "_PERSONA_ID_",
      "tag": undefined,
    },
  },
  Object {
    "node": Object {
      "$$typeof": Symbol(machinat.element),
      "props": Object {},
      "type": "bar",
    },
    "path": "$::1",
    "type": "unit",
    "value": Object {
      "message": Object {
        "metadata": undefined,
        "quick_replies": undefined,
        "text": "bar",
      },
      "messaging_type": undefined,
      "notification_type": undefined,
      "persona_id": "_PERSONA_ID_",
      "tag": undefined,
    },
  },
]
`;

exports[`Dialog with quickReplies match snapshot 1`] = `
Array [
  Object {
    "node": "foo",
    "path": "$::0",
    "type": "unit",
    "value": Object {
      "message": Object {
        "text": "foo",
      },
      "messaging_type": undefined,
      "notification_type": undefined,
      "persona_id": undefined,
      "tag": undefined,
    },
  },
  Object {
    "node": Object {
      "$$typeof": Symbol(machinat.element),
      "props": Object {},
      "type": "bar",
    },
    "path": "$::1",
    "type": "unit",
    "value": Object {
      "message": Object {
        "metadata": undefined,
        "quick_replies": Array [
          "_QUICK_REPLY_RENDERED_",
          "_QUICK_REPLY_RENDERED_",
          "_QUICK_REPLY_RENDERED_",
          "_QUICK_REPLY_RENDERED_",
        ],
        "text": "bar",
      },
      "messaging_type": undefined,
      "notification_type": undefined,
      "persona_id": undefined,
      "tag": undefined,
    },
  },
]
`;

exports[`Dialog with type and tag match snapshot 1`] = `
Array [
  Object {
    "node": "foo",
    "path": "$::0",
    "type": "unit",
    "value": Object {
      "message": Object {
        "text": "foo",
      },
      "messaging_type": "MESSAGE_TAG",
      "notification_type": undefined,
      "persona_id": undefined,
      "tag": "COMMUNITY_ALERT",
    },
  },
  Object {
    "node": Object {
      "$$typeof": Symbol(machinat.element),
      "props": Object {},
      "type": "bar",
    },
    "path": "$::1",
    "type": "unit",
    "value": Object {
      "message": Object {
        "metadata": undefined,
        "quick_replies": undefined,
        "text": "bar",
      },
      "messaging_type": "MESSAGE_TAG",
      "notification_type": undefined,
      "persona_id": undefined,
      "tag": "COMMUNITY_ALERT",
    },
  },
]
`;

exports[`Dialog with type match snapshot 1`] = `
Array [
  Object {
    "node": "foo",
    "path": "$::0",
    "type": "unit",
    "value": Object {
      "message": Object {
        "text": "foo",
      },
      "messaging_type": "UPDATE",
      "notification_type": undefined,
      "persona_id": undefined,
      "tag": undefined,
    },
  },
  Object {
    "node": Object {
      "$$typeof": Symbol(machinat.element),
      "props": Object {},
      "type": "bar",
    },
    "path": "$::1",
    "type": "unit",
    "value": Object {
      "message": Object {
        "metadata": undefined,
        "quick_replies": undefined,
        "text": "bar",
      },
      "messaging_type": "UPDATE",
      "notification_type": undefined,
      "persona_id": undefined,
      "tag": undefined,
    },
  },
]
`;
